File: src/main/java/org/algohub/engine/type/InternalTestCase.java
package org.algohub.engine.type;

import com.google.common.base.Preconditions;
import org.algohub.engine.judge.Function;
import org.algohub.engine.judge.Problem;

public class InternalTestCase {
    private final Object[] input;
    private final Object output;
    private final boolean returnsVoid;

    public InternalTestCase(final Problem.TestCase testCase, final Function function) {
        final Function.Parameter[] parameters = function.getParameters();
        Preconditions.checkArgument(parameters.length == testCase.getInput().size());

        input = new Object[testCase.getInput().size()];
        for (int i = 0; i < input.length; ++i) {
            input[i] = Deserializer.fromJson(parameters[i].getType(), testCase.getInput().get(i));
        }

        returnsVoid = IntermediateType.VOID.equals(function.getReturnStatement().getType().getValue());

        if (returnsVoid) {
            this.output = Deserializer.fromJson(parameters[0].getType(), testCase.getOutput());
        } else {
            this.output = Deserializer.fromJson(function.getReturnStatement().getType(), testCase.getOutput());
        }
    }

    public Object[] getInput() {
        return input;
    }

    public Object getOutput() {
        return output;
    }

    public boolean returnsVoid() {
        return returnsVoid;
    }
}

File: src/main/kotlin/com/jalgoarena/judge/InternalTestCase.kt
package com.jalgoarena.judge

import com.google.common.base.Preconditions
import com.jalgoarena.ObjectMapperInstance
import org.slf4j.LoggerFactory
import java.io.IOException

internal class InternalTestCase(testCase: Problem.TestCase, function: Function) {

    private val LOG = LoggerFactory.getLogger(InternalTestCase::class.java)

    val input: Array<Any?>
    val output: Any?
    val returnsVoid: Boolean

    init {
        val parameters = function.parameters
        Preconditions.checkArgument(parameters.size == testCase.input.size())

        try {
            input = arrayOfNulls(testCase.input.size())
            for (i in input.indices) {
                input[i] = deserialize(
                        testCase.input.get(i).toString(),
                        Class.forName(parameters[i].type)
                )
            }

            returnsVoid = "void" == function.returnStatement.type

            val outputType =
                    if (returnsVoid) Class.forName(parameters[0].type)
                    else Class.forName(function.returnStatement.type)

            this.output = deserialize(
                    testCase.output.toString(),
                    outputType
            )
        } catch (e: ClassNotFoundException) {
            LOG.error(e.message, e)
            throw IllegalArgumentException(e.message)
        } catch (e: IOException) {
            LOG.error(e.message, e)
            throw IllegalArgumentException(e.message)
        }

    }

    private fun deserialize(content: String, type: Class<*>): Any? {
        return ObjectMapperInstance.INSTANCE.readValue(content, type)
    }
}
